### PR TITLE
Break sequence expressions 

### LINKED ISSUE #2240: Odd linebreak for sequence of assignment expressions with ternaries
```js
(a = b ? c : 'lllllllllllllllllllllll'),
  (a = b ? c : 'lllllllllllllllllllllll'),
  (a = b ? c : 'lllllllllllllllllllllll'),
  (a = b ? c : 'lllllllllllllllllllllll'),
  (a = b ? c : 'lllllllllllllllllllllll')
```
[formats as](https://prettier.github.io/prettier/#%7B%22content%22%3A%22(a%20%3D%20b%20%3F%20c%20%3A%20'lllllllllllllllllllllll')%2C%5Cn%20%20(a%20%3D%20b%20%3F%20c%20%3A%20'lllllllllllllllllllllll')%2C%5Cn%20%20(a%20%3D%20b%20%3F%20c%20%3A%20'lllllllllllllllllllllll')%2C%5Cn%20%20(a%20%3D%20b%20%3F%20c%20%3A%20'lllllllllllllllllllllll')%2C%5Cn%20%20(a%20%3D%20b%20%3F%20c%20%3A%20'lllllllllllllllllllllll')%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D)

```js
(a = b ? c : "lllllllllllllllllllllll"), (a = b
  ? c
  : "lllllllllllllllllllllll"), (a = b ? c : "lllllllllllllllllllllll"), (a = b
  ? c
  : "lllllllllllllllllllllll"), (a = b ? c : "lllllllllllllllllllllll");
```

which seems... very wrong.

### PR DESCRIPTION
**Pull request by bakkot**
Fixes #2240.

This is split into two commits so you can look at just the second to get a better idea of how this affects output.

~~This isn't going to pass CI, because of an un-idempotency:~~ (Edit: just removed the test...)

```js
function sequenceExpression() {
  return (
    // Reason for a
    a
  ), b
}
```
prints as
```js
function sequenceExpression() {
  return (
    // Reason for a
    a,
    b
  );
}
```
prints as
```js
function sequenceExpression() {
  return (
    // Reason for a
    a, b
  );
}
```

I can't for the life of me figure out where the linebreak after the `a` is coming from, and `--debug-print-doc` seems broken. (@vjeux, help?) Otherwise this is probably ready to go.

### CODE REVIEW

**Comment by vjeux** [comment]
Feel free to get rid of the failing test, I don't think we've actually seen this one in real code.

File: src/printer.js [hunk 1]
```diff
@@ -1124,25 +1136,26 @@ function genericPrintNoParens(path, options, print, args) {
...
+      if (
+        parent.type === "ExpressionStatement" ||
+        parent.type === "ForStatement"
+      ) {
+        // For ExpressionStatements and for-loop heads, which are among
+        // the few places a SequenceExpression appears unparenthesized, we want
+        // to indent expressions after the first.
+        const parts = [];
+        path.each(p => {
+          if (p.getName() === 0) {
```
Comment (vjeux) [inline src/printer.js]: This is super weird to use getName. Can you create a local variable i and increment it within the each?

File: tests/flow/optional/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -349,7 +349,8 @@ function foo(x?: number, ...y: Array<string>): [?number, Array<string>] {
 }
 
 foo(); // OK
-foo(123), foo(123, "hello"); // OK // OK
+foo(123), // OK
+  foo(123, "hello"); // OK
```
Comment (vjeux) [inline tests/flow/optional/__snapshots__/jsfmt.spec.js.snap]: That actually fixes this case :)

File: tests/sequence_break/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -2,12 +2,75 @@
...
+    aLongIdentifierName,
+    aLongIdentifierName,
+    aLongIdentifierName
+  )
+);
+for (
+  aLongIdentifierName = 0,
+    aLongIdentifierName = 0,
+    aLongIdentifierName = 0,
+    aLongIdentifierName = 0;
```
Comment (vjeux) [inline tests/sequence_break/__snapshots__/jsfmt.spec.js.snap]: Did we indent there before? Or is it new with this pr

Comment (vjeux) [inline tests/sequence_break/__snapshots__/jsfmt.spec.js.snap]: Also what happens if we have a var before, do we indent it?

File: src/printer.js (hunk 1 above)
Comment (bakkot) [inline src/printer.js]: Could do, though this is actually a [documented use](https://github.com/prettier/prettier/blob/a0cadad2220a7843516de5f1ca44f10f0bccc5ad/src/fast-path.js#L93-L94) of getName. Are you sure you want it changed?

File: tests/sequence_break/__snapshots__/jsfmt.spec.js.snap (hunk 3 above)
Comment (bakkot) [inline tests/sequence_break/__snapshots__/jsfmt.spec.js.snap]: We just didn't break at all previously - you can see in the first previous commit [what things looked like](https://github.com/prettier/prettier/pull/2388/commits/b18b6846f30c89cb93b985008e7eba7ff1216953#diff-e1baa6054f27ebf4c3ea72fb5aea5bc5R35) before the change.

And yup, we [indent declarations](https://prettier.io/playground/#%7B%22content%22%3A%22for%20(var%20aLongIdentifierName%20%3D%200%2C%20aLongIdentifierName%20%3D%200%2C%20aLongIdentifierName%20%3D%200%2C%20aLongIdentifierName%20%3D%200%3B%20test%3B%20update)%20%7B%7D%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Afalse%7D%7D) here.

**Comment by bakkot** [comment]
> Feel free to get rid of the failing test, I don't think we've actually seen this one in real code.

Done, though it hurts my soul.

I'd really like to get debug-print-doc working so I can fix it properly, but I've created a [tracking issue](https://github.com/prettier/prettier/issues/2392) for now.

**Comment by vjeux** [comment]
Can you open an issue for debug-print-doc as well? It should be fixed. I would recommend using bisect to figure out when it broke.

**Pull request merged**